test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ModelBuilder/RecursiveComplexTypesTests.cs

```diff
@@ -0,0 +1,489 @@
+            IEdmModel model = builder.GetEdmModel();
+
+            var field =
+                model.SchemaElements.First(e => e.Name == typeof(RecursiveComplexTypesTests_Field).Name)
+                as EdmComplexType;
+
+            var subFieldsProperty =
+                field.Properties().Single(p => p.Name == "SubFields") as EdmStructuralProperty;
+
+            Assert.Equal(
+                typeof(RecursiveComplexTypesTests_Field).Name,
+                subFieldsProperty.Type.AsCollection()?.ElementType().AsComplex()?.ComplexDefinition().Name);
+        }
+
+        [Fact]
+        public void CanBuildModelWithIndirectRecursiveReference()
+        {
+            var builder = new ODataConventionModelBuilder();
+            builder.EntitySet<RecursiveComplexTypesTests_JustCustomer>("justcustomer");
+            IEdmModel model = builder.GetEdmModel();
+
+            string customerTypeName = typeof(RecursiveComplexTypesTests_Customer).Name;
+
+            var customer = model.SchemaElements.First(e => e.Name == customerTypeName) as EdmComplexType;
+
+            var accountsProperty =
+                customer.Properties().Single(p => p.Name == "Accounts") as EdmStructuralProperty;
+
+            string accountTypeName = typeof(RecursiveComplexTypesTests_Account).Name;
+
+            Assert.Equal(
+                accountTypeName,
+                accountsProperty.Type.AsCollection()?.ElementType().AsComplex()?.ComplexDefinition().Name);
+
+            var account = model.SchemaElements.First(e => e.Name == accountTypeName) as EdmComplexType;
+            var ownerProperty = account.Properties().Single(p => p.Name == "Owner") as EdmStructuralProperty;
+
+            Assert.Equal(customerTypeName, ownerProperty.Type.AsComplex()?.ComplexDefinition().Name);
+        }
+
+        [Fact]
+        public void CanBuildModelWithCollectionReferenceViaInheritance()
+        {
+            var builder = new ODataConventionModelBuilder();
+            builder.EntitySet<RecursiveComplexTypesTests_JustHomeDirectory>("justhomedirectory");
+            IEdmModel model = builder.GetEdmModel();
+
+            string fileTypeName = typeof(RecursiveComplexTypesTests_File).Name;
+            string directoryTypeName = typeof(RecursiveComplexTypesTests_Directory).Name;
+
+            var file = model.SchemaElements.First(e => e.Name == fileTypeName) as EdmComplexType;
+            var directory = model.SchemaElements.First(e => e.Name == directoryTypeName) as EdmComplexType;
+
+            Assert.Equal(fileTypeName, directory.BaseComplexType()?.Name);
+
+            var filesProperty = directory.Properties().Single(p => p.Name == "Files") as EdmStructuralProperty;
+
+            Assert.Equal(
+                fileTypeName,
+                filesProperty.Type.AsCollection()?.ElementType().AsComplex()?.ComplexDefinition().Name);
+        }
+    }
+
+    public class RecursiveComplexTypesE2ETests : WebHostTestBase
```

Comment:
>RecursiveComplexTypesE2ETests [](start = 17, length = 29)

I'd like to rename the test class name same as the file name.

Reply:
Done